### PR TITLE
mpi: autogen attr api

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Use python scripts to autogen attr api functons.

The functions and any customizations are specified in `src/binding/c/attr_api.txt`.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/binding/c/attr_api.txt [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+/*
+    Notes for C:
+        Even though the 'attribute_val' argument is declared as 'void *', it is
+        really the address of a void pointer (i.e., a 'void **').  Using
+        a 'void *', however, is more in keeping with C idiom and allows the
+        pointer to be passed without additional casts.
+
+    .N Fortran
+        The 'attribute_val' in Fortran is a pointer to a Fortran integer, not
+        a pointer to a 'void *'.
```
Comment (shintaro-iwasaki) [inline src/binding/c/attr_api.txt:214]: Do we need to add this comment?
https://github.com/pmodels/mpich/blob/main/src/mpi/attr/win_get_attr.c#L108-L113

File: src/binding/c/attr_api.txt [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    In C and C++, an attribute value is a pointer ('void *'); in Fortran, it is an
+    address-sized integer.
+
+    If an attribute is already present, the delete function (specified when the
+    corresponding keyval was created) will be called.
+*/
+
+MPI_Type_create_keyval:
+    .desc: Create an attribute keyval for MPI datatypes
+    .extra: AttrErrReturn
```
Comment (shintaro-iwasaki) [inline src/binding/c/attr_api.txt:104]: Minor comment: can we make these flags a bit more readable? This makes reviewers easy to review the PR (especially to check which keyword affects which part). Regarding this PR:
1.  `AttrErrReturn`
   If it only affects the comment, how about `DocAttrErrReturn` or using a different `.xxx` (say `.doc: AttrErrReturn `)?
2. `ignore_revoke`
   Is there a rule for capitalization? For example, should it be  `IgnoreRevoke`? I appreciate if the keyword contains more information (e.g., `IgnoreRevokedComm`).

**Review by shintaro-iwasaki — COMMENTED** [review]
Thank you very much for you change! If all tests pass, I think it's okay regarding the last `binding/c: autogen attr api` commit.

File: src/binding/c/attr_api.txt (hunk 2 above)
Comment (hzhou) [reply src/binding/c/attr_api.txt:104]: `AttrErrReturn` are the exact phrase defined in `maint/docnotes` and in the output as `.N AttrErrReturn`.

`.doc: ` is a good idea!

I don't really like camel case. How about `ignore_revoked_comm`? 

But it is still cryptic as you may still don't understand what does ignore_revoked_comm means. In the end, all these directives are cryptic -- either you know what it means and what you want or don't. The alternative is to directly write custom "error_check" code block. The reason we opt for directives is for brevity and easier maintenance in the long run (when maintainer understands what these directives means or knows how to look them up). So I am hesitate for the attempt to make the word more descriptive.

File: src/binding/c/attr_api.txt (hunk 1 above)
Comment (hzhou) [reply src/binding/c/attr_api.txt:214]: Will add.

File: src/binding/c/attr_api.txt (hunk 2 above)
Comment (shintaro-iwasaki) [reply src/binding/c/attr_api.txt:104]: > `.doc`

That's great!

> But it is still cryptic as you may still don't understand what does ignore_revoked_comm means.

I understand, but that does not mean we can keep `ignore_revoke`. I believe at least we should include domains (in this case, maybe "error" and/or "comm") in the keyword. I don't think making such a keyword short makes maintenance easier. A non-maintainer (like reviewers of the potential future PRs) cannot understand it.

I think `ignore_revoked_comm` is much better.  Perhaps putting `error` would make it more readable.

Comment (hzhou) [reply src/binding/c/attr_api.txt:104]: I'll work on it (add `.doc` and `ignore_revoked_comm`)

**Comment by hzhou** [comment]
> AttrErrReturn
If it only affects the comment, how about DocAttrErrReturn or using a different .xxx (say .doc: AttrErrReturn )?

@shintaro-iwasaki I updated the python scripts in PR #4953 now that we will use 
```
    .docnotes: AttrErrReturn, ...
```
for any docnotes that listed in `maint/docnotes`. There is no need to define or document what notes are allowed or what do they mean, they will be output as is and the definitions are in `maint/docnotes`.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/binding/c/attr_api.txt (hunk 2 above)
Comment (hzhou) [reply src/binding/c/attr_api.txt]: Added `.docnotes:` and renamed to `ignore_revoked_comm`

File: src/binding/c/attr_api.txt (hunk 1 above)
Comment (hzhou) [reply src/binding/c/attr_api.txt:219]: Added.

**Review by shintaro-iwasaki — APPROVED** [review]
@hzhou Thanks a lot! It looks good to me.

**Pull request merged**